Readme.md
# Jac - just another configuration lib

[![Gem](https://img.shields.io/gem/v/jac.svg)](https://rubygems.org/gems/jac)
[![Coverage Status](https://img.shields.io/codeclimate/coverage/vizor-games/jac.svg)](https://codeclimate.com/github/vizor-games/jac)
[![Code Climate](https://codeclimate.com/github/vizor-games/jac/badges/gpa.svg)](https://codeclimate.com/github/vizor-games/jac)
[![Build Status](https://travis-ci.org/vizor-games/jac.svg?branch=master)](https://travis-ci.org/vizor-games/jac)

## Installation and usage

To start using jac you need to add 

```ruby
gem 'jac', '~> 0.0.5'
```

to your `Gemfile` and load configuration from default paths (`jac.yml`, `jac.user.yml`) relative to working dir:

```ruby
require 'jac'
profile = %w[any profile combination]
Jac::Configuration.load(profile) # => OpenStruct
```

or to load custom set of files:

```ruby
require 'jac'
profile = %w[any profile combination]
Jac::Configuration.load(profile, files: %w[example/config/base.yml example/config/custom.yml]) # => OpenStruct
```

## Features

Configuration is loaded from well formed YAML streams.
Each document expected to be key-value mapping where
keys a `profile` names and values is a profile content.
Profile itself is key-value mapping too. Except reserved
key names (i.e. `extends`) each key in profile is a
configuration field. For example following yaml document

```yml

foo:
  bar: 42
qoo:
  bar: 32

```

represents description of two profiles, `foo` and `qoo`,
where field `bar` is set to `42` and `32` respectively.

Profile can be constructed using combination of other profiles
for example having `debug` and `release` profiles for testing
and production. And having `remote` and `local` profiles for
building on local or remote machine. We cant get `debug,local`,
`debug,remote`, `release,local` and `release,remote` profiles.
Each of such profiles is a result of merging values of listed
profiles. When merging profile with another configuration
resolver overwrites existing fields. For example if `debug`
and `local` for some reason have same field. In profile
`debug,local` value from `debug` profile will be overwritten
with value from `local` profile.

### Extending profiles

One profile can `extend` another. Or any amount of other
profiles. To specify this one should use `extends` field
like that

```yml

base:
  application_name: my-awesome-app
  port: 80

version:
  version_name: 0.0.0
  version_code: 42

debug:
  extends: [base, version]
  port: 9292
```

In this example `debug` will receive following fields:

```yml
application_name: my-awesome-app  # from base profile
port: 9292                        # from debug profile
version_name: 0.0.0               # from version profile
version_code: 42                  # from version profile
```

### Merging multiple configuration files

Configuration can be loaded from multiple YAML documents.
Before resolve requested profile all described profiles
are merged down together. Having sequence of files like
`.application.yml`, `.application.user.yml` with following content

```yml
# .application.yml
base:
  user: deployer

debug:
  extends: base
  # ... other values
```

```yml
# .application.user.yml
base:
  user: developer
```

We'll get `user` field overwritten with value from
`.application.user.yml`. And only after that construction
of resulting profile will begin (for example `debug`)

### String evaluation

Configuration resolver comes with powerful yet dangerous
feature: it allows evaluate strings as ruby expressions
like this:

```yml
foo:
  build_time: "#{Time.now}" # Will be evaluated at configuration resolving step
```

Configuration values are available to and can be referenced with `c`:

```yml
base:
  application_name: my-awesome-app
debug:
  extends: base
  server_name: "#{c.application_name}-debug"   # yields to my-awesome-app-debug
release:
  extends: base
  server_name: "#{c.application_name}-release" # yields to my-awesome-app-release
```

_local_ values can be referenced through self, e.g.:

```yml
default:
  project:
    network: fb
    server_name: "#{self['network']}-srv" # => 'fb-srv'
```

All strings evaluated **after** profile is constructed thus
you don't need to have declared values in current profile
but be ready to get `nil`.

### Merging values

By default if one value have multiple defenitions it will be overwritten by
topmost value. Except several cases where Jac handles value resolution
specialy

#### Merging hashes

Hashes inside profile are recurseively merged automatically. This rule works
for profile extensions and value redefenitions too.

Example:

```yml
base:
  servers:
    release: 'http://release.com'

debug:
  extends: base
  servers:                      # will contain {'debug' => 'https://debug.com', 'release' => 'https://release.com'}
    debug: 'http://debug.com'

```

#### Merging sets

Sets allowed to be merged with `nil`s and any instance of `Enumerable`.
Merge result is always new `Set` instance.

Example: 
```yml
release:
  extends:
    - no_rtti
    - no_debug
  flags: !set {} # empty set
no_rtti:
  flags:
    - '-fno-rtti'
no_debug:
  flags:
    - '-DNDEBUG'
```

Resulting profile will have `Set('-fno-rtti', '-DNDEBUG')` in `release profile`

## Generic profiles

Same result as shown above can be achieved with generic profiles. Generic profile
is a profile which name is regex (i.e. contained in `/.../`):

```yml
base:
  application_name: 'my-awesome-app'
/(release|debug)/: # Profile name is a regex, with single capture (1)
  extends: base
  server_name: "#{c.application_name}-#{c.captures[1]}"  # yields  my-awesome-app-release or  my-awesome-app-debug
```

If profile name matches multiple generic profiles it not defined
which profile will be used.

>  If running on Ruby 2.4+ you can use named captures in generic profiles.
> Named captures will be stored in `c.named_captures` as `Hash`. 

## Implicit profiles

`jac` has two implicit profiles which automaticly aplied to any configuration
it produces: `^base` and `^top` _(note `^` at the begining of profile name)_.

If for some reason you need to provide base values for *any* profile you use
you should define them in `^base` profile. And if you need always apply some 
values onto any configuration you receive you should define them in `^top`
profile.

```yaml
^base:
   java: 1.7 # Use java 1.7 by default

^top:
   Xmx: 512M # Override any Xmx value if present
```

Using this example we can get `default` profile containing

```
java: 1.7
Xmx: 512M
```

## License

jac is licensed under the MIT licence. Please see the [LICENSE](LICENSE) for more information.


Changelog.md
# Version 0.0.5

* [Fixed] Merger#merge! fails to call `resolve_values` properly on JRuby 9.2.0

# Version 0.0.4

* [#3](https://github.com/vizor-games/jac/issues/3) Allow reference local values inside string evaluations
* [#4](https://github.com/vizor-games/jac/issues/4) Add implicit `^top` and `^base` profiles
* Allow reference local values inside string evaluations at top level keys
* Allow reference profile name at config evaluation
* Profile keys shadow object methods in string evaluations

# Version 0.0.3

* Allow load configuration from custom locations

# Version 0.0.1 

* First release


spec/spec_helper.rb
require 'tmpdir'
require 'powerpack/string'
require 'simplecov'

SimpleCov.start
require_relative './helpers/nested'

# @param [String] path relative to lib folder
# @return [String] full path
def lib(path)
  File.join(__dir__, '..', 'lib', path)
end

RSpec.configure do |c|
  c.include Helpers::Nested, nested: true
end


spec/helpers/nested.rb
module Helpers
  module Nested
    def nested(levels, opts = {})
      opts.inject({}) do |acc, elem|
        key, value = elem
        acc.update(key => levels == 0 ? value : nested(levels - 1, opts))
      end
    end
  end
end


spec/jac/merger_spec.rb
require_relative lib('jac/merger')
require_relative lib('jac/value')

describe Configuration::Merger do
  let(:merger) { Configuration::Merger.new }

  describe '#merge', nested: true do
    it 'merges hashes recursievly' do
      a = Values.of(foo: { qoo: 2 })
      b = Values.of(foo: { bar: 3 })
      c = { foo: { bar: 3, qoo: 2 } }
      expect(merger.merge(a, b)._value).to eq(c)
    end

    it 'merges deep hashes' do
      a = Values.of(nested(2, a: 1, b: 2))
      b = Values.of(nested(2, a: 1, b: 3))
      c = nested(2, a: 1, b: 3)
      expect(merger.merge(a, b)._value).to eq(c)
    end

    it 'does not enter into arrays' do
      a = Values.of(a: [{ b: 3 }])
      b = Values.of(a: [{ b: 4 }])
      expect(merger.merge(a, b)._value).to eq(b._value)
    end

    it 'merges sets' do
      a = Values.of(a: Set.new(1..3))
      b = Values.of(a: Set.new(4..6))
      c = Values.of(a: Set.new(1..6))
      expect(merger.merge(a, b)._value).to eq(c._value)
    end

    context 'when one value is nil but other is set' do
      it 'returns set' do
        a = Values.of(a: nil)
        b = Values.of(a: Set.new(1..3))
        expect(merger.merge(a, b)._value).to eq(b._value)
        expect(merger.merge(b, a)._value).to eq(b._value)
      end
    end
  end
end


lib/jac/value.rb
require 'set'

module Jac
  module Values
    class << self
      def of(o, attributes = {})
        klass = case o
                when Array
                  ArrayValue
                when Hash
                  HashValue
                when Set
                  SetValue
                when String
                  StringValue
                else
                  ObjectValue
                end

        klass.new(o, attributes)
      end
    end

    # Value class keeps track of where value
    # comes from, its context, etc.
    class Value
      def initialize(o, attributes)
        @object = o
        @attributes = OpenStruct.new(attributes)
      end

      def _value
        @object
      end

      def method_missing(meth, *args, &block)
        @object.send(meth, *args, &block)
      end
    end

    class ArrayValue < Value
      def _value
        @object.map(&:_value)
      end
    end

    class HashValue < Value
      def _value
        @object.inject({}) do |acc, elem|
          k, v = elem.map(&:_value)
          acc.update(k => v)
        end
      end
    end

    class StringValue < Value
    end

    class SetValue < Value
      def _value
        Set.new(@object.map(&:_value))
      end
    end

    class ObjectValue < Value
    end
  end
end


lib/jac/merger.rb
require 'set'

module Jac
  module Configuration
    # Merges two hashes with following value resolve strategy:
    # * When having both `that: Hash` and `other: Hash` for same key will merge
    #   them with same strategy and return result
    # * When having Set and Enumerable will join two sets into
    #   one
    # * When having Set and nil will return Set
    # * Return `other` value in all other cases
    class Merger
      # Returns a new hash with base and overrides merged recursively.
      # @param [Hash] base values
      # @param [Hash] other values
      # @return [Hash] updated hash
      def merge(base, other)
        merge!(base.dup, other)
      end

      # Returns a new hash with base and overrides merged recursively. Updates
      # receiver
      # @param [Hash] base values
      # @param [Hash] other values
      # @return [Hash] updated base hash
      def merge!(base, other)
        # base.merge!(other, &method(:resolve_values))
        base.merge!(other) do |key, base_value, other_value|
          resolve_values(key, base_value, other_value)
        end
      end

      # @param [Object] _key
      # @param [Object] base
      # @param [Object] other
      # @return [Object] resolved value
      def resolve_values(_key, base, other)
        if base.nil? && other.nil?
          nil
        elsif to_hash?(base, other)
          merge(base, other)
        elsif to_set?(base, other)
          Set.new(base) + Set.new(other)
        else
          other
        end
      end

      private

      def to_hash?(base, other)
        base.is_a?(Hash) && other.is_a?(Hash)
      end

      def to_set?(base, other)
        (base.is_a?(Set) && set_like?(other)) || (other.is_a?(Set) && set_like?(base))
      end

      # rubocop: disable Naming/AccessorMethodName
      def set_like?(value)
        value.is_a?(Enumerable) || value.nil?
      end
      # rubocop: enable Naming/AccessorMethodName
    end
  end
end


lib/jac/parser.rb
require 'psych'
require 'set'

require_relative 'value'

module Jac
  # Custom YAML parsing routines
  module Parser
    # List of features supported in Psych by version
    FEATURES = {
      locations: '>=3.0.2' # Psych::Node contains start_line, end_line,
      # start_column, end_column
    }.freeze

    class << self
      def features(psych_version = Psych::VERSION)
        FEATURES
          .select { |_k, v| Versions.match(psych_version, v) }
          .keys
      end
    end
    # Cutstom Yaml AST visitor
    # @see Psych::Visitors::ToRuby
    # While standard Psych visitor converts sets to `{ value => nil }` mappings
    # we need explicitly convert those mappings to ruby [Set]
    class VisitorToRuby < Psych::Visitors::ToRuby
      attr_accessor :stream_name

      # Uses standard Psych visitor to convert mapping to ruby object
      # except `!set` case. Here we convert mapping to [Set].
      # @param [Psych::Nodes::Mapping] o YAML AST node
      # @return [Object] parsed ruby object
      def visit_Psych_Nodes_Mapping(o) # rubocop:disable Naming/MethodName
        case o.tag
        when '!set', 'tag:yaml.org,2002:set'
          visit_set(o)
        else # fallback to default implementation
          super(o)
        end
      end

      def accept(o)
        attributes = { source: stream_name }
        if Parser.features.include?(:locations)
          attributes[:start_line] = o.start_line
          attributes[:end_line] = o.end_line
          attributes[:start_column] = o.start_column
          attributes[:end_column] = o.end_column
        end
        Values.of(super(o), attributes)
      end

      private

      def visit_set(o)
        set = Set.new
        # Update anchor
        @st[o.anchor] = set if o.anchor
        o.children.each_slice(2) do |k, _v|
          set << accept(k)
        end
        set
      end
    end
  end
end


lib/jac/versions.rb
# Utilitary module for comparing versions
module Versions
  class << self
    OPERATORS = {
      '>=' => :>=,
      '<=' => :<=,
      '=' => :==
    }.freeze
    # Tries to match provided version with target rule
    # If version matches, will return version value. If not, returns nil.
    # Target rule can be described in following walys
    # * '=x.y.z' - exact match with version
    # * '>=x.y.z' - requires minimum version as x.y.z
    # * '<=x.y.z' - limits wersion with x.y.z
    #
    # @param version [String] verstion string to match
    # @param target_rule [String] rule describing version requirement
    # @return [String] version string or nil, if version don't match
    #  rule
    def match(version, target_rule)
      op, v = target_rule.match(/^(=|>=|<=)(.+)$/) do |m|
        [OPERATORS[m[1]], m[2]]
      end

      raise('Version format does not match: ' + target_rule) unless op
      return version if version.send(op, v)
    end
  end
end


lib/jac/configuration.rb
require 'yaml'
require 'ostruct'

module Jac
  # Configuration is loaded from well formed YAML streams.
  # Each document expected to be key-value mapping where
  # keys a `profile` names and values is a profile content.
  # Profile itself is key-value mapping too. Except reserved
  # key names (i.e. `extends`) each key in profile is a
  # configuration field. For example following yaml document
  #
  # ```yml
  #
  #  foo:
  #    bar: 42
  #  qoo:
  #    bar: 32
  #
  # ```
  #
  # represents description of two profiles, `foo` and `qoo`,
  # where field `bar` is set to `42` and `32` respectively.
  #
  # Profile can be constructed using combination of other profiles
  # for example having `debug` and `release` profiles for testing
  # and production. And having `remote` and `local` profiles for
  # building on local or remote machine. We cant get `debug,local`,
  # `debug,remote`, `release,local` and `release,remote` profiles.
  # Each of such profiles is a result of merging values of listed
  # profiles. When merging profile with another configuration
  # resolver overwrites existing fields. For example if `debug`
  # and `local` for some reason have same field. In profile
  # `debug,local` value from `debug` profile will be overwritten
  # with value from `local` profile.
  #
  # ## Extending profiles
  #
  # One profile can `extend` another. Or any amount of other
  # profiles. To specify this one should use `extends` field
  # like that
  #
  # ```yml
  #
  # base:
  #   application_name: my-awesome-app
  #   port: 80
  #
  # version:
  #   version_name: 0.0.0
  #   version_code: 42
  #
  # debug:
  #   extends: [base, version]
  #   port: 9292
  # ```
  #
  # In this example `debug` will receive following fields:
  #
  # ```yml
  # application_name: my-awesome-app  # from base profile
  # port: 9292                        # from debug profile
  # version_name: 0.0.0               # from version profile
  # version_code: 42                  # from version profile
  # ```
  #
  # ## Merging multiple configuration files
  #
  # Configuration can be loaded from multiple YAML documents.
  # Before resolve requested profile all described profiles
  # are merged down together. Having sequence of files like
  # `.application.yml`, `.application.user.yml` with following content
  #
  # ```yml
  # # .application.yml
  # base:
  #   user: deployer
  #
  # debug:
  #   extends: base
  #   # ... other values
  # ```
  #
  # ```yml
  # # .application.user.yml
  # base:
  #  user: developer
  # ```
  #
  # We'll get `user` field overwritten with value from
  # `.application.user.yml`. And only after that construction
  # of resulting profile will begin (for example `debug`)
  #
  # ## String evaluation
  #
  # Configuration resolver comes with powerful yet dangerous
  # feature: it allows evaluate strings as ruby expressions
  # like this:
  #
  # ```yml
  # foo:
  #   build_time: "#{Time.now}" # Will be evaluated at configuration resolving step
  # ```
  #
  # Configuration values are available to and can be referenced with `c`:
  #
  # ```yml
  # base:
  #   application_name: my-awesome-app
  # debug:
  #   extends: base
  #   server_name: "#{c.application_name}-debug"   # yields to my-awesome-app-debug
  # release:
  #   extends: base
  #   server_name: "#{c.application_name}-release" # yields to my-awesome-app-release
  # ```
  #
  # All strings evaluated **after** profile is constructed thus
  # you don't need to have declared values in current profile
  # but be ready to get `nil`.
  #
  # ## Merging values
  #
  # By default if one value have multiple defenitions it will be overwritten by
  # topmost value. Except several cases where Jac handles value resolution
  # specialy
  #
  # ### Merging hashes
  #
  # Hashes inside profile are recurseively merged automaticly. This rule works
  # for profile extensions and value redefenitions too.
  #
  # Example:
  #
  # ```yml
  # base:
  #   servers:
  #     release: http://release.com
  #
  # debug:
  #   extends: base
  #     debug: http://debug.com
  #
  # ```
  #
  # ### Merging sets
  #
  # Sets allowed to be merged with `nil`s and any instance of `Enumerable`.
  # Merge result is always new `Set` instance.
  # ```yml
  # release:
  #   extends:
  #     - no_rtti
  #     - no_debug
  #   flags: !set {} # empty set
  # no_rtti:
  #   flags:
  #     - -fno-rtti
  # no_debug:
  #   flags:
  #     - -DNDEBUG
  # ```
  #
  # Resulting profile will have `-fno-rtti, -DNDEBUG` in `release profile`
  # ## Generic profiles
  #
  # Same result as shown above can be achieved with generic profiles. Generic profile
  # is a profile which name is regex (i.e. contained in `/.../`):
  #
  # ```
  # base:
  #   application_name: my-awesome-app
  # /(release|debug)/: # Profile name is a regex, with single capture (1)
  #   extends: base
  #   server_name: "#{c.application_name}-#{c.captures[1]}"  # yields  my-awesome-app-release or  my-awesome-app-debug
  # ```
  #
  # If profile name matches multiple generic profiles it not defined
  # which profile will be used.
  module Configuration
    # Name of top level profile. ^top profile is implicit
    # it automaticly merged on top of resulting configuration
    TOP_PROFILE_NAME = '^top'.freeze
    # Name of base level profile. ^base profile is implicit
    # resulting configuration automaticly merged on top of it.
    BASE_PROFILE_NAME = '^base'.freeze
    # Any configuration set always contains `default` profile
    # which is loaded when no profile requested.
    DEFAULT_PROFILE_NAME = 'default'.freeze
    # List of default profiles. Having this list we can easily create default
    # configuration.
    BASIC_PROFILES = [
      TOP_PROFILE_NAME,
      BASE_PROFILE_NAME,
      DEFAULT_PROFILE_NAME
    ].freeze
    # Key where all inherited profiles listed
    EXTENDS_KEY = 'extends'.freeze

    # Reads and evaluates configuration for given set of streams
    # and profile
    class ConfigurationReader
      attr_reader :merger
      # Creates configuration reader
      # @param [Array] streams of pairs containing YAML document and provided
      #   name for this stream
      def initialize(streams)
        @streams = streams
        @merger = Merger.new
      end

      # Parses all streams and resolves requested profile
      # @param [Array] profile list of profile names to be merged
      # @return [OpenStruct] instance which contains all resolved profile fields
      def read(*profile)
        result = @streams
                 .flat_map { |stream, _name| read_stream(stream) }
                 .inject(default_configuration) { |acc, elem| update(acc, elem) }
        # Keep original profile name
        original_profile = profile
        # Add implicit profiles
        profile =
          [Configuration::BASE_PROFILE_NAME, profile, Configuration::TOP_PROFILE_NAME].flatten
        OpenStruct.new(evaluate(resolve(profile, result).merge('profile' => original_profile)))
      end

      private

      # Creates empty configuration object.
      def default_configuration
        Configuration::BASIC_PROFILES.inject({}) { |acc, elem| acc.update(elem => {}) }
      end

      def read_stream(stream)
        # Each stream consists of one or more documents
        YAML.parse_stream(stream).children.flat_map do |document|
          # Will use separate visitor per YAML document.
          visitor = Jac::Parser::VisitorToRuby.create
          # Expecting document to be single mapping
          profile_mapping = document.children.first
          raise(ArgumentError, 'Mapping expected') unless profile_mapping.is_a? Psych::Nodes::Mapping
          # Then mapping should be expanded to (key, value) pairs. Because yaml overwrites
          # values for duplicated keys. This is not desired behaviour. We need to merge
          # such entries
          profile_mapping
            .children
            .each_slice(2)
            .map { |k, v| { visitor.accept(k) => visitor.accept(v) } }
        end
      end

      def update(config, config_part)
        config_part.each do |profile, values|
          profile_values = config[profile]
          unless profile_values
            profile_values = {}
            config[profile] = profile_values
          end
          merge!(profile_values, values)
        end

        config
      end

      # Merges two hash structures using following rules
      # @param [Hash] base value mappings
      # @param [Hash] values ovverides.
      # @return [Hash] merged profile
      def merge!(base, values)
        merger.merge!(base, values)
      end

      def resolve(profile, config)
        ProfileResolver.new(config).resolve(profile)
      end

      def evaluate(resolved_profile)
        ConfigurationEvaluator.evaluate(resolved_profile)
      end
    end

    # Describes profile resolving strategy
    class ProfileResolver
      attr_reader :config, :merger

      def initialize(config)
        @config = config
        @merger = Merger.new
      end

      def resolve(profile, resolved = [])
        profile.inject({}) do |acc, elem|
          if resolved.include?(elem)
            msg = 'Cyclic dependency found ' + (resolved + [elem]).join(' -> ')
            raise(ArgumentError, msg)
          end
          profile_values = find_profile(elem)
          # Find all parent_profiles
          extends = parent_profiles(profile_values)
          # We can do not check extends. Empty profile returns {}
          # Inherited values goes first
          inherited = merger.merge(acc, resolve(extends, resolved + [elem]))
          merger.merge(inherited, profile_values)
        end
      end

      # Fetches list of parent profile
      # @param [Hash] values current profile values
      # @return [Array] of profile names
      def parent_profiles(values)
        extends = *values[Configuration::EXTENDS_KEY] || []
        [Configuration::TOP_PROFILE_NAME, Configuration::BASE_PROFILE_NAME].each do |implicit|
          raise(ArgumentError, "`#{implicit}` is not allowed here") if extends.include?(implicit)
        end
        extends
      end

      def find_profile(profile_name)
        return config[profile_name] if config.key?(profile_name)
        # First and last chars are '/'
        match, matched_profile = find_generic_profile(profile_name)
        # Generating profile
        return generate_profile(match, matched_profile, profile_name) if match
        raise(ArgumentError, 'No such profile ' + profile_name)
      end

      def generate_profile(match, matched_profile, profile_name)
        gen_profile = {}
        gen_profile['captures'] = match.captures if match.captures
        if match.respond_to?(:named_captures) && match.named_captures
          gen_profile['named_captures'] = match.named_captures
        end
        gen_profile.merge!(config[matched_profile])

        config[profile_name] = gen_profile
      end

      # @todo print warning if other matching generic profiles found
      def find_generic_profile(profile_name)
        generic_profiles(config)
          .detect do |profile, regex|
            m = regex.match(profile_name)
            break [m, profile] if m
          end
      end

      def generic_profiles(config)
        # Create generic profiles if missing
        @generic_profiles ||= config
                              .keys
                              .select { |k| k[0] == '/' && k[-1] == '/' }
                              .map { |k| [k, Regexp.new(k[1..-2])] }
      end
    end

    # Proxy class for getting actual values
    # when referencing profile inside evaluated
    # expressions
    class EvaluationContext
      def initialize(hash, evaluator)
        @evaluator = evaluator
        hash.each_key do |key|
          exp = <<-EVAL.strip_indent
          def #{key}
            @evaluator.evaluate("#{key}")
          end
          EVAL
          eval(exp, binding)
        end
      end
    end

    # Evaluates all strings inside resolved profile
    # object
    class ConfigurationEvaluator
      def initialize(src_object)
        @object = src_object
        @evaluated = {}
        # Initialize context object with root
        @context = EvaluationContext.new(@object, self)
      end

      def evaluate(key)
        return @evaluated[key] if @evaluated.key? key
        @evaluated[key] = evaluate_deep(@object[key])
      end

      def c
        @context
      end

      alias config c
      alias conf c
      alias cfg c

      def evaluate_values
        @ctx_object = @context
        @object.each_key { |k| evaluate(k) }
        # Cleanup accidentally created values (when referencing missing values)
        @evaluated.delete_if { |k, _v| !@object.key?(k) }
      end

      private

      def get_binding(object)
        binding
      end

      def evaluate_deep(object) # rubocop:disable Metrics/MethodLength
        case object
        when String
          eval_string(object)
        when Array
          contextual(object) do
            object.map { |e| evaluate_deep(e) }
          end
        when Hash
          # Evaluating values only by convention
          contextual(object) do
            object.inject({}) { |acc, elem| acc.update(elem.first => evaluate_deep(elem.last)) }
          end
        else
          object
        end
      end

      def eval_string(o)
        evaluated = /#\{.+?\}/.match(o) do
          eval('"' + o + '"', get_binding(@ctx_object))
        end

        evaluated || o
      end

      def respond_to_missing?(meth, args, &block)
        @ctx_object && @ctx_object.respond_to?(meth, args, &block)
      end

      def method_missing(meth, *args, &block)
        # rubocop inspection hack
        return super unless respond_to_missing?(meth, args, &block)
        @ctx_object.send(meth, *args, &block)
      end

      # Keeps track of current context objects
      # This trick allows us to reference local variables
      # using `self` calls.
      # @param ctx [Object] object to use in current evaluation
      # @param &block [Proc] evaluation logic
      # @return [Object] result of evaluation
      def contextual(ctx)
        memo = @ctx_object
        @ctx_object = ctx
        result = yield
        @ctx_object = memo

        result
      end

      class << self
        def evaluate(o)
          ConfigurationEvaluator.new(o).evaluate_values
        end
      end
    end

    # List of files where configuration can be placed
    # * `jac.yml` - expected to be main configuration file.
    # Usually it placed under version control.
    # * `jac.user.yml` - user defined overrides for main
    # configuration, sensitive data which can't be placed
    # under version control.
    # * `jac.override.yml` - final overrides.
    CONFIGURATION_FILES = %w[jac.yml jac.user.yml jac.override.yml].freeze

    class << self
      # Generates configuration object for given profile
      # and list of streams with YAML document
      # @param profile [Array] list of profile names to merge
      # @param streams [Array] list of YAML documents and their
      # names to read
      # @return [OpenStruct] instance which contains all resolved profile fields
      def read(profile, *streams)
        profile = [Configuration::DEFAULT_PROFILE_NAME] if profile.empty?
        ConfigurationReader.new(streams).read(*profile)
      end

      # Read configuration from configuration files.
      # @praram [String or Array] profile which should be loaded
      # @param [Array] files filenames to load
      # @param [String] dir base directory path for provided files may be nil
      # @return [OpenStruct] resolved profile values
      def load(profile, files: CONFIGURATION_FILES, dir: nil)
        # Read all known files
        streams = files
                  .map { |f| [dir ? File.join(dir, f) : f, f] }
                  .select { |path, _name| File.exist?(path) }
                  .map { |path, name| [IO.read(path), name] }
        read(profile, *streams)
      end
    end
  end
end
